### PR TITLE
[do not merge] [stdlib] Improved pointers

### PR DESCRIPTION
**Pull request by tayloraswift**
[Implementation of SE-184](https://github.com/apple/swift-evolution/pull/741)

### CODE REVIEW

**Review by CodaFi — CHANGES_REQUESTED** [review]
Blocking this for visibility: 

Throughout this diff there's a bunch of little whitespace changes.  As a general policy, we like to have diffs as clean as possible.  Could you go through and scrub the whitespace changes?

File: stdlib/public/core/Sequence.swift
Comment (xwu) [inline stdlib/public/core/Sequence.swift]: Here and elsewhere, can you make sure to follow prevailing stdlib convention, specifically:
* Place access modifiers on the same line as `func`
* Use no space after the opening parenthesis
* Use no space before colon and one space after colon (unless stating conformances or inheritance, in which case use one space before and after)
* Use two-space tabs and 80-character columns
* Place braces on the same line as `else`, `for`, etc. (including on the same line as the function return type); place `else` on the same line as `guard`
* Use no spaces before and after `..<`

File: stdlib/public/core/StringCore.swift
Comment (xwu) [inline stdlib/public/core/StringCore.swift]: Here and below, please restore the original spacing, which indents two spaces when breaking after an opening parenthesis.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Here and below, please follow prevailing style above; besides the style notes above (`public` on same line as `init`, space after `:`, `{` on same line as arguments), set internal members without prefixing `self` and do not align `_end =` with spaces.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Can't this be simply `%  else:` after eliminating the `%  end` above?

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Besides other style notes above (including not using additional spaces in `raw =`), please indent only two spaces here.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Here, stdlib style is to line-break after the opening parenthesis and increase indent by two spaces on the next line.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: In comments, stdlib style is to capitalize and punctuate like a sentence.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Ditto here about stdlib style and parentheses; I'll not repeat comments above about the styling of `guard...else` statements, which also apply here.

**Review by xwu — COMMENTED** [review]
Excited about these improvements! I've left some style notes below.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Where type annotations are not necessary, stdlib style is to leave them off. Also, stdlib style is for camel-casing even local variables (i.e. `dstStart`).

**Comment by atrick** [comment]
I know, the whitespace thing is terribly annoying! I used to open/close a file to fix the whitespace then commit that separately before editing the file. Pure whitespace changes are also discouraged in Swift, but it's probably fine if it's limited to a file that you're already making large changes to.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (atrick) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Please add a scary warning comment to UnsafeMutableBufferPointer.deinitialize(). Something like:

Warning: All buffer elements must be initialized before calling this. Note that invoking `initialize(from:)` may not initialize all elements. Deinitializing part of the buffer must be done using the `UnsafeMutablePointer` API: `baseAddress.deinitialize(count: n)`

**Comment by kelvin13** [comment]
I found out how to disable whitespace stripping in Atom but I have no idea how to restore the whitespace that was already there before Atom stripped it since it’s all mixed with the actual changes.

**Comment by CodaFi** [comment]
For the files with whitespace only changes `git checkout master path/to/file` will reset them.  For the files with actual code changes you may have to do it by hand

**Comment by kelvin13** [comment]
i only saved files i made changes to, which was about 20 or so because of function and argument renaming to make the tests compile

**Comment by kelvin13** [comment]
does it look better?

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: done

File: stdlib/public/core/Sequence.swift
Comment (xwu) [inline stdlib/public/core/Sequence.swift]: One more stdlib-specific style note: the prevailing convention is to put the closing parenthesis, deindented, next to the return arrow: `) ->`.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Style nit: missing spaces here; should be `S : Sequence`.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Nit: I think this now all fits on one line :)

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Style nit: This one's tricky, but `public func` should be one line, then probably to balance out the opening `(`, you could line break and deindent before `)`--although I'm not sure about this one; check elsewhere in the codebase.

Below, where you've got `body(${Self}<T>(` and `count: ` on a separate line, line return after the opening parenthesis and indent two spaces, then put everything else on that line.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Style nits: space after `switch`; `default` instead of `case _` (which isn't used in stdlib unless i'm mistaken).

More generally, though, isn't this more idiomatically just:

```swift
guard let position = _position, let end = _end else { return 0 }
return end - position
```

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Style nit: I believe stdlib always explicitly marks internal members as `internal`.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: A good place for a multiline string literal?

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: Nit: the original line breaking here was consistent with prevailing stdlib convention.

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: Nit: ditto here.

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: Nit: move `{` to previous line and `}` into a separate line.

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: Nit: as before, `) ->` is the prevailing style.

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: Nit: ditto here.

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: ...and here.

**Review by xwu — COMMENTED** [review]
Some more style nits (sorry!).

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: i don’t think i ever modified this part of the code, the diff is probably because a method near it got moved

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: is there a stdlib convention for this? my own convention is to always put `guard` and `if` comma statements on their own lines

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: again, wasn’t me

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: again, wasn’t me

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: i thought dummy blocks weren’t to be formatted like “real” function bodies, like `a.map{ $0.foo }` instead of 

```swift
a.map 
{
    $0.foo 
}
```

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Fair; since the diff now points to you, it could stand a little stylistic tidying though :)

File: stdlib/public/core/UnsafeRawPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeRawPointer.swift.gyb]: All function bodies have the format

```swift
public func foo() {
  fatalError()
}
```

See: https://github.com/apple/swift/blob/f6f3ed0fe7e3c0d8d6af85bb238e039041ac0d42/validation-test/stdlib/CollectionDiagnostics.swift#L15

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (xwu) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: In the stdlib, if the whole thing fits, you'll see everything in one line:

```swift
guard let foo = _foo, let bar = _bar else { return }
```

(Which is not to say, of course, that your style is inferior in any way.)

**Review by CodaFi — APPROVED** [review]

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (natecook1000) [inline stdlib/public/core/UnsafePointer.swift.gyb]: Using `unavailable` in these changes will break existing code. Should this be something like `@available(swift, deprecated: 4.1, obsoleted: 5, message: "...")`? Do we have the machinery in place for those versions yet?

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (natecook1000) [inline stdlib/public/core/UnsafePointer.swift.gyb]: Please update docs when the API changes—in this case, the name of the method is now `intialize(repeating:count)` and the first parameter name has changed from `newValue` to `repeatedValue`.

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (natecook1000) [inline stdlib/public/core/UnsafePointer.swift.gyb]: Default parameter values need to be specified in the docs for the parameter: "The default value is `1`". Noted here and 4–5 other places below.

File: stdlib/public/core/UnsafeBufferPointer.swift.gyb
Comment (natecook1000) [inline stdlib/public/core/UnsafeBufferPointer.swift.gyb]: Please make sure any new public APIs have documentation before this gets merged—thanks!

**Review by natecook1000 — CHANGES_REQUESTED** [review]
Looking great! I've added a few notes about where we need to make sure we're keeping documentation in line with the changes.

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafePointer.swift.gyb]: as explained in the [proposal](https://github.com/kelvin13/swift-evolution/blob/9d305ced7b82b9cf5854b55b3f0d08952853d046/proposals/0184-improved-pointers.md#source-compatibility) we are preferring a hard `unavailable` over a soft deprecation.

**Comment by gottesmm** [comment]
@kelvin13 Please make sure to rebase the branch on master before merging!

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafePointer.swift.gyb]: done

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (kelvin13) [inline stdlib/public/core/UnsafePointer.swift.gyb]: dont

**Comment by kelvin13** [comment]
The implementation is now up to speed with the [latest version](https://github.com/kelvin13/swift-evolution/blob/9d305ced7b82b9cf5854b55b3f0d08952853d046/proposals/0184-improved-pointers.md#source-compatibility) of the draft proposal

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (natecook1000) [inline stdlib/public/core/UnsafePointer.swift.gyb]: I see that, but using `unavailable` causes source breakage of Swift 4 code, not just code that is migrated to Swift 5.

**Review by natecook1000 — APPROVED** [review]

File: stdlib/public/core/UnsafePointer.swift.gyb
Comment (natecook1000) [reply stdlib/public/core/UnsafePointer.swift.gyb]: Thanks for taking care of these!

**Pull request closed without merging**
